test/WebJobs.Script.Tests/MetricsEventManagerTests.cs

```diff
@@ -0,0 +1,357 @@
+                        }
+                        else
+                        {
+                            hashes.Add(lastEvent.InvocationId);
+                        }
+
+                        var minEventsExpected = Math.Floor((double)lastEvent.ExecutionTimeSpan / (double)MinimumLongRunningDurationInMs) - 1;
+                        var maxEventsExpected = Math.Ceiling((double)lastEvent.ExecutionTimeSpan / (double)MinimumLongRunningDurationInMs) + 1;
+                        // We should see atleast one InProgress event if it takes more than 5 seconds
+                        if (lastEvent.ExecutionTimeSpan >= MinimumLongRunningDurationInMs
+                            && (relatedEventIds.Count < minEventsExpected
+                            || relatedEventIds.Count > maxEventsExpected))
+                        {
+                            invalidElement = lastEvent;
+                            errorMessage = string.Format("Long running function doesn't contain expected number of Etw events. Minimum:{0} Maximum:{1} Actual:{2}", minEventsExpected, maxEventsExpected, relatedEventIds.Count);
+                            return false;
+                        }
+
+                        foreach (var relatedEventId in relatedEventIds)
+                        {
+                            // Mark all related events as processed
+                            functionValidationTrackerList[relatedEventId].HasBeenProcessed = true;
+                        }
+                    }
+                    else if (functionExecutionArgs.ExecutionStage == ExecutionStage.Finished.ToString())
+                    {
+                        functionValidationTrackerList[currentIndex].HasBeenProcessed = true;
+                        hashes.Add(functionExecutionArgs.InvocationId);
+                    }
+                }
+            }
+
+            var unprocessedEvents = functionValidationTrackerList.Where(e => !e.HasBeenProcessed).ToList();
+            if (unprocessedEvents.Count > 0)
+            {
+                invalidElement = unprocessedEvents[0].EventArguments;
+                errorMessage = string.Format("There are unprocessed events: {0}", SerializeFunctionExecutionEventArguments(unprocessedEvents.Select(e => e.EventArguments).ToList()));
+                return false;
+            }
+
+            if (hashes.Count != noOfFuncExecutions)
+            {
+                invalidElement = unprocessedEvents[0].EventArguments;
+                errorMessage = string.Format("No of finished events does not match with number of function executions: Expected:{0} Actual:{1}", noOfFuncExecutions, hashes.Count);
+                return false;
+            }
+
+            return true;
+        }
+
+        private static WebHostMetricsLogger CreateWebHostMetricsLoggerInstance(List<FunctionExecutionEventArguments> functionExecutionEventArgumentsList)
+        {
+            var metricsEventGenerator = new Mock<IMetricsEventGenerator>();
```

Comment:
You don't need Moq for this - recommend simply implementing your interface to collect the events